modules/generator/processor/servicegraphs/servicegraphs.go

```diff
@@ -280,6 +273,72 @@ func (p *Processor) upsertPeerNode(e *store.Edge, spanAttr []*v1_common.KeyValue
 	}
 }
 
+// upsertDatabaseRequest handles the logic of adding a database edge on the
+// graph.  If we have a db.name or db.system attribute, we assume this is a
+// database request.  The name of the edge is determined by the following
+// order:
+//
+//	if we have a peer.service, use it as the database ServerService
+//	if we have a server.address, use it as the database ServerService
+//	if we have a db.name, use it as the database ServerService, which is the backwards-compatible behavior
+func (p *Processor) upsertDatabaseRequest(e *store.Edge, resourceAttr []*v1_common.KeyValue, span *v1_trace.Span) {
+	var (
+		isDatabase bool
+
+		// The fallback database name
+		dbName string
+	)
+
+	// Check for db.name first.  The dbName is set initially to maintain backwards compatbility.
+	if name, ok := processor_util.FindAttributeValue(string(semconv.DBNameKey), resourceAttr, span.Attributes); ok {
+		dbName = name
+		isDatabase = true
+	}
+
+	// Check for db.system only if we don't have db.name above
+	if !isDatabase {
+		if _, ok := processor_util.FindAttributeValue(string(semconv.DBSystemKey), resourceAttr, span.Attributes); ok {
+			isDatabase = true
+		}
+	}
+
+	// If neither db.system nor db.name are present, we can't determine if this is a database request
+	if !isDatabase {
+		return
+	}
+	e.ConnectionType = store.Database
+	e.ServerLatencySec = spanDurationSec(span)
+
+	// Set the service name by order of precedence
+
+	// Check for peer.service
+	if name, ok := processor_util.FindAttributeValue(string(semconv.PeerServiceKey), resourceAttr, span.Attributes); ok {
+		e.ServerService = name
+		return
+	}
+
+	// Check for server.address
+	if name, ok := processor_util.FindAttributeValue(string(semconv.ServerAddressKey), resourceAttr, span.Attributes); ok {
+		e.ServerService = name
+		return
+	}
+
+	// Check for network.peer.address and network.peer.port.  Use port if it is present.
+	if host, ok := processor_util.FindAttributeValue(string(semconv.NetworkPeerAddressKey), resourceAttr, span.Attributes); ok {
+		if port, ok := processor_util.FindAttributeValue(string(semconv.NetworkPeerPortKey), resourceAttr, span.Attributes); ok {
+			e.ServerService = fmt.Sprintf("%s:%s", host, port)
```

Comment:
little concerned about cardinality here. it should be fine b/c it should accurately reflect real databases in an org.

we had similar issues before and @mapno submitted this.

https://github.com/grafana/tempo/pull/2505

my guess is that it's fine, but we should keep an eye out for cardinality jumps in prod as we roll this out.

Reply:
I thought about the carnality here too, but I think because this is server side information, we should be okay since I expect server name/port combos to be significantly less than the ephemeral ports in use by the client.  Good thing to watch out for when we roll for sure.